processing/heartbeat_injector.go

```diff
@@ -0,0 +1,127 @@
+package processing
+
+// DCSO FEVER
+// Copyright (c) 2020, DCSO GmbH
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"regexp"
+	"time"
+
+	"github.com/DCSO/fever/types"
+	"github.com/DCSO/fever/util"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	// match 24 hour local time string, separated by colon
+	injectTimeRegex     = regexp.MustCompile(`^(([01][0-9])|(2[0-3])):[0-5][0-9]$`)
+	injectTimeCheckTick = 40 * time.Second
+)
+
+// HeartbeatInjector regularly adds a date-based pseudo-event to the forwarded
+// event stream.
+type HeartbeatInjector struct {
+	SensorID       string
+	Times          []string
+	CloseChan      chan bool
+	Logger         *log.Entry
+	ForwardHandler Handler
+}
+
+// MakeHeartbeatInjector creates a new HeartbeatInjector.
+func MakeHeartbeatInjector(forwardHandler Handler, injectTimes []string) (*HeartbeatInjector, error) {
+	sensorID, err := util.GetSensorID()
+	if err != nil {
+		return nil, err
+	}
+	for _, v := range injectTimes {
+		if !injectTimeRegex.Match([]byte(v)) {
+			return nil, fmt.Errorf("invalid time specification in heartbeat injector config: '%s'", v)
+		}
+	}
+	a := &HeartbeatInjector{
+		ForwardHandler: forwardHandler,
+		Logger: log.WithFields(log.Fields{
+			"domain": "heartbeat_injector",
+		}),
+		Times:     injectTimes,
+		CloseChan: make(chan bool),
+		SensorID:  sensorID,
+	}
+	return a, nil
+}
+
+func makeHeartbeatEvent() types.Entry {
+	now := time.Now()
+	e := types.Entry{
+		SrcIP:     "192.0.2.1",
+		SrcPort:   int64(rand.Intn(60000) + 1025),
```

Comment:
Configurable heartbeatEvent by configurable template is a usecase for a later PR maybe? Not sure if this is doable by go vanilla features.

Reply:
Not really needed right now but one might think of actually using a Go template for the even, right. There is a templating engine in the standard library, which I have already used in a different project. Later.